### PR TITLE
[RLlib] ConvTranspose2D module

### PR DESCRIPTION
**Pull request by sven1977**
A ConvTranspose2D Module for torch (to be used for image generating models).
- plus test case.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/models/tests/test_convtranspose2d_stack.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    """Tests our ConvTranspose2D Stack modules/layers."""
+
+    def test_convtranspose2d_stack(self):
+        """Tests, whether the conv2d stack can be trained to predict an image.
+        """
+        batch_size = 128
+        input_size = 1
+        module = ConvTranspose2DStack(input_size=input_size)
+        optim = torch.optim.Adam(module.parameters(), lr=0.0001)
+        img = cv2.imread("../../tests/data/images/obstacle_tower.png")
```
Comment (michaelzhiluo) [inline rllib/models/tests/test_convtranspose2d_stack.py:23]: This is a bit janky.

File: rllib/models/tests/test_convtranspose2d_stack.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        # Make channels first.
+        img = np.transpose(img, (2, 0, 1))
```
Comment (michaelzhiluo) [inline rllib/models/tests/test_convtranspose2d_stack.py:25]: Should be a generic pixel preprocessing function

File: rllib/models/tests/test_convtranspose2d_stack.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        imgs = (imgs / 255.0) * 2.0 - 1.0
+        imgs = torch.from_numpy(imgs)
+        init_loss = loss = None
+        for _ in range(10):
+            # Random inputs.
+            inputs = torch.from_numpy(
+                np.random.normal(0.0, 1.0, (batch_size, input_size))).float()
+            distribution = module(inputs)
+            # Construct a loss.
+            loss = -torch.mean(distribution.log_prob(imgs))
```
Comment (michaelzhiluo) [inline rllib/models/tests/test_convtranspose2d_stack.py:39]: Are there preexisting agents that use conv transpose stack?

Comment (michaelzhiluo) [reply rllib/models/tests/test_convtranspose2d_stack.py:39]: Then you could probably simplify this code

File: rllib/models/torch/modules/convtranspose2d_stack.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            # format (channels first).
+            Reshape([-1, in_channels, 1, 1]),
+        ]
+        for i, (_, kernel, stride) in enumerate(filters):
+            out_channels = filters[i + 1][0] if i < len(filters) - 1 else \
+                output_shape[0]
+            conv_transp = nn.ConvTranspose2d(
+                in_channels, out_channels, kernel, stride)
+            # Apply initializer.
+            initializer(conv_transp.weight)
```
Comment (michaelzhiluo) [inline rllib/models/torch/modules/convtranspose2d_stack.py:57]: Change initalizer by overriding the reset_parameters method: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/linear.py#L48-L52

File: rllib/models/torch/modules/reshape.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from typing import List
+
+from ray.rllib.utils.framework import try_import_torch
+
+torch, nn = try_import_torch()
+
+
+class Reshape(nn.Module):
```
Comment (michaelzhiluo) [inline rllib/models/torch/modules/reshape.py:8]: Can we add this to the file where the SlimFC stuff is located instead?

**Review by michaelzhiluo — CHANGES_REQUESTED** [review]

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 1 above)
Comment (sven1977) [reply rllib/models/tests/test_convtranspose2d_stack.py:23]: It's a test case :)

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/tests/test_convtranspose2d_stack.py:25]: +1, do we already have anything like that somewhere?

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 3 above)
Comment (sven1977) [reply rllib/models/tests/test_convtranspose2d_stack.py:39]: Dreamer, no? :) Other than that, nothing yet, to my knowledge. I just made up something quick here to test the model + some loss.

File: rllib/models/torch/modules/convtranspose2d_stack.py (hunk 4 above)
Comment (sven1977) [reply rllib/models/torch/modules/convtranspose2d_stack.py:57]: +1

File: rllib/models/torch/modules/reshape.py (hunk 5 above)
Comment (sven1977) [reply rllib/models/torch/modules/reshape.py:8]: +1

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 2 above)
Comment (michaelzhiluo) [reply rllib/models/tests/test_convtranspose2d_stack.py:25]: Not sure... If you look in Dreamer dmc2gym in examples/envs, there is similar preprocessing

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 1 above)
Comment (michaelzhiluo) [reply rllib/models/tests/test_convtranspose2d_stack.py:23]: Haha ok i'll leave it as it is

File: rllib/models/tests/test_convtranspose2d_stack.py (hunk 3 above)
Comment (michaelzhiluo) [reply rllib/models/tests/test_convtranspose2d_stack.py:39]: Aight wfm

File: rllib/models/torch/modules/reshape.py (hunk 5 above)
Comment (sven1977) [reply rllib/models/torch/modules/reshape.py:8]: done

File: rllib/models/torch/modules/convtranspose2d_stack.py (hunk 4 above)
Comment (sven1977) [reply rllib/models/torch/modules/convtranspose2d_stack.py:57]: I like the above solution better. Allows us to have all initialization logic within one module (rllib/models/torch/modules/convtranspose2d_stack.py), not having to create an extra sub-class of torch.nn.ConvTranspose2d (like in Dreamer model, where the initialization logic is hard-coded).

Comment (michaelzhiluo) [reply rllib/models/torch/modules/convtranspose2d_stack.py:57]: Aight this wfm

File: rllib/models/torch/misc.py [hunk 6]
```diff
@@ -138,3 +139,15 @@ def forward(self, x):
         out = torch.cat(
             [x, self.log_std.unsqueeze(0).repeat([len(x), 1])], axis=1)
         return out
+
```
Comment (sven1977) [inline rllib/models/torch/misc.py:142]: Moved this to model/torch/misc

**Review by michaelzhiluo — APPROVED** [review]

File: rllib/agents/dreamer/dreamer_model.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 from typing import Any, List, Tuple
+from ray.rllib.models.torch.modules.reshape import Reshape
```
Comment (seminumber) [inline rllib/agents/dreamer/dreamer_model.py:3]: This line crashes. The previous dialog indicates that you seem to move it to ray.rllib.models.torch.misc .
